core/vm/tosca_integration.go

```diff
@@ -116,6 +116,9 @@ func (in *EVMInterpreter) Step(state *InterpreterState) {
 		state.ReturnData = res
 	} else if err != nil {
 		state.Status = Failed
+	} else if len(state.Contract.Code) == 0 {
+		// No contract code is treated as a stop, to avoid infinite loops
+		state.Status = Stopped
```

Comment:
It looks to me like this condition is actually covering up an implementation error in the `run` function. In line [189](https://github.com/Fantom-foundation/go-ethereum-sonic/blob/2400937cc3b1c370ea8735d7af86eac621e4eb70/core/vm/interpreter.go#L189) of the interpreter, there is an early exist if the code is empty. However, I think there is a bug: it is intended to "behave" like the interpretation of a STOP instruction, but it misses to set the `state.Error` to the `errStopToken`.

I think it would be cleaner to
- make or that the effect of the early exit is equivalent to a STOP instruction, or
- remove the early exit

The first option is probably the better one, since the early exit is part of the Ethereum code we want to preserve as a reference ([here](https://github.com/ethereum/go-ethereum/blob/08e6bdb550712503873fb2a138b30132cc36c481/core/vm/interpreter.go#L176-L179)). But adding the status update in the condition's body is sound, since we added the state object in the first place. This is an additional requirement we imposed to implement the StepN function.

Reply:
Right, we are introducing the state object to the `run` function therefore we have to set the `errStopToken` in the state.

I originally tried to keep the changes to the tosca_integration.go file, as this will make updating the ethereum version easier, but this introduces a special case in an unexpected place.